### ISSUE TITLE
WIREFRAME:  Advanced Filtering [3]

### ISSUE THREAD

**Issue by catrobson** [fabric8io/fabric8-ux#194]
As a developer, I'd like to be able to filter the Work Item list using only text syntax to make advanced filters.

Systems that do this:
* Github
* Jira

Ideas from the team:
Max A: i would suggest treat the field in the combo to be an "advanced query" string and parse it to make it do the RightThing(tm)
Michael K: I would think that when I click on a icon/button on the filter bar, the "labels" are collapsing to an expression that can be edited. Like in the file managers that use breadcrumps. Hitting CTRL-L collapses the breadcrumbs to an expression/path.

**Comment by catrobson** [comment]
Samir has been working on design improvements fro advanced filtering.  Please add links to your work as you progress.

**Comment by sdash-redhat** [comment]
First draft created is available here [https://redhat.invisionapp.com/share/QABI3D5MX#/screens](https://redhat.invisionapp.com/share/QABI3D5MX#/screens)

Got initial feedback from Catherine. Have started working on creating examples based on a simplified version shortly. Meanwhile please share your feedback on the current draft version.

This initial draft version focuses on a few things to do the heavy lifting like -- 

- Filtering based on nested parameters
- Advanced  filtering options like AND / OR relationships among parameters
- Optionally providing an approach to customise most used/popular filters up front etc. 

Basically there 2 types of information we might deal with across the application  -- the data itself (the backlog items, files, staus, authors, etc. ) and the second is the set of meta data or the secondary layer of info. like the tags . The tags can have multi levels of relationships.  The proposed interaction was trying to do the heavy lifting across multi-levels of information hierarchy among the metadata and can be customised to fit in the simplest use case also where you just need one linear hierarchy of information and just one level of meta-data. 

The "preferred filters" are placed with a goal to be accessed upfront. There can tweaks like use of just icons, that are commonly understood... or replacing radio-boxes with toggle states. Keeping "preferred filters" as separate block will help in well position them in responsive layout in smaller resolutions. They can appear as next to the text field however it has to be upfront and should not be in one level down for access.

**Comment by sdash-redhat** [comment]
@catrobson : I am adding your feedback for reference with my comments inline [] : 

Thanks for putting together this first draft!  I took a look through it.  At first pass, it feels complex to understand for the end user with the different components.  
**[SAMIR: agree that too many components can be simplified. will look into that aspect. may be we can remove the customisation section out ]**

One thing that might help make it more consumable to me, and possibly others, is if you add a real example of how the Planner team might use this multi-faceted filtering capability to create a real filter they need for their backlog, and show a wireframe with realistic data based on that.  Its hard to tell whether it would pull together and make more sense once there was real data in it.
**[SAMIR: will be providing some real life example in next version ... lets say using WI or build status context]**

Another observation I had was that It felt like having two "areas" to the filter - the box w/ has advanced features, and the right hand preferred filters as another interaction separate from the filter dropdown was a bit confusing.  What do you think about having everything fairly co-located?
**[SAMIR: yeah..fundamentally they are two different components, visually we can put them together with closer proximity ... let's explore]**

Also consider that I don't think we would use this type of filtering component on something like codebases where we have a pretty simple list.  I expect we could stay with a "simple filter" from PatternFly for this, whereas the Planner list view is much more complex and needs a much richer interaction for filtering.  With the planner view, you could say that the left hand panel (Iterations, Backlog) is a "filter" and that we could build out that to include more rich filters.  I could see us having "All" (all work items), "Backlog" (currently statically defined filter, but we could let people configure what backlog means to them), "My Named Filter 1", "My Named filter 2",etc. before the iterations listing - or pulling the iterations into a "view" of that left hand panel that is slightly different from the work item filtered list views that are system-created or self-created.

One feature that I know Todd wants us to have that I did not see in this is a way for me to "Save" my filter as a "quick filter" for future use.  JIRA and gmail have this type of a feature if you want to see how others do it.
**[SAMIR: yes..this is an user need i missed out ..will add]**

**Comment by sdash-redhat** [comment]
Attached the draft 0.2 to invision for the advance filtering. 
Please review and share your thoughts. https://redhat.invisionapp.com/share/GUBPEPBCW

(Also you can get at google slide https://docs.google.com/a/redhat.com/presentation/d/1fZf3VZbvujz3tcGweuIz1phlp0MvVLFKUlwWoZDx6ZI/edit?usp=sharing)

@catrobson @muruGanesan, @Essjaysee , @mindreeper2420 , @bdellasc, @joshuawilson , @aslakknutsen

**Comment by sdash-redhat** [comment]
There is a call with Matt Carrano tonight, who is working on a PF solution on advance filter for BPM. Will try to get his views /thoughts on this approch and if there is any other clues I get from any other members who attend the call on the same to ensure, we cover the all critical user-needs covered.

**Comment by sdash-redhat** [comment]
Had a discussion with last evening on PF advance filter gaps and the typical needs/use-cases with Matt Carrano, Liz Clayton, Ming Zhong,  Matthew Stevens. Some thoughts from the call: 
1. The proposed solution does fill the gaps pointed out. 
2. In some usecases the users may need some advanced features more than that is offered in quick filter, but they may not need complex query building. - though this is an assumption and further usecases are being looked into. 
In such scenario, the proposed solution approach can be used till where it opens the popup to modify the filters with relationships..
3. The query builder can be plugged into wherever required and can be exlored with another story. 

Looking at these here are my takeaways: 

1.  The proposed solution can be split into 2 stories - 1 . for a filter that can handle both simple and medium complexities of queries (i.e. the prosed solution excluding the query builder)  2. the later will be another story that can specifically define the query builder part  and both of them can be 2 different modules. 

Basically the current proposed solution combines both the needs. 

The first part will be sufficient to close the current story  and the later will help creating a new PF widget for query building. 

Share your thoughts/views.

**Comment by sdash-redhat** [comment]
I have updated the wireframe and have uploaded to the new version 0.3 to Invision. 
Please review and share your thoughts. 

**https://redhat.invisionapp.com/d/main#/projects/prototypes/11063451**

Source at Google drive: https://docs.google.com/a/redhat.com/presentation/d/1R5DUlOgILODh4NUsrZzO3fhREhHliHOWyFx1QDim82U/edit?usp=sharing

**Comment by catrobson** [comment]
Sprint 131c:  Focus on smaller screen sizes and how all the elements will fit into the filter header bar with changes that Brian is introducing with #401.

**Comment by sdash-redhat** [comment]
Updated to show how it will behave in lower resolutions 
Also how it will fit in to the new layout designed by Brian [#401](https://github.com/fabric8io/fabric8-ux/issues/401).

Check the updated version 0.4 at 
[https://redhat.invisionapp.com/share/X6BYQRBVN](https://redhat.invisionapp.com/share/X6BYQRBVN)

Primarily review these slides #42, #43, #44

@catrobson @muruGanesan, @Essjaysee , @mindreeper2420 , @bdellasc, @joshuawilson , @aslakknutsen
